pkg/rangemap/rm.go

```diff
@@ -0,0 +1,95 @@
+package rangemap
+
+import (
+	"errors"
+	"github.com/anujga/dstk/pkg/core"
+	"github.com/google/btree"
+)
+
+var (
+	ErrRangeOverlaps = errors.New("range overlaps")
+	ErrKeyAbsent = errors.New("key absent")
+)
+
+type DummyRange struct {
+	start core.KeyT
+}
+
+func (ki *DummyRange) Start() core.KeyT {
+	return ki.start
+}
+
+func (ki *DummyRange) End() core.KeyT {
+	return nil
+}
+
+type RangeMap struct {
+	root *btree.BTree
+}
+
+func New(degree int) *RangeMap {
+	return &RangeMap{root: btree.New(degree)}
+}
+
+func (rm *RangeMap) getLessOrEqual(item *rangeItem) *rangeItem {
+	var itemInTree *rangeItem
+	rm.root.DescendLessOrEqual(item, func(i btree.Item) bool {
+		itemInTree = i.(*rangeItem)
+		return false
+	})
+	return itemInTree
+}
+
+func (rm *RangeMap) Get(key core.KeyT) (Range, error) {
+	item, err := NewRange(&DummyRange{start: key})
+	if err != nil {
+		return nil, err
+	}
+	pred := rm.getLessOrEqual(item)
+	if pred.contains(key) {
+		return pred.Range, nil
+	}
+	return nil, ErrKeyAbsent
+}
+
+func (rm *RangeMap) Put(rng Range) error {
+	item, err := NewRange(rng)
+	if err != nil {
+		return err
+	}
+	pred := rm.getLessOrEqual(item)
+	if pred != nil && !pred.preceeds(item) {
+		return ErrRangeOverlaps
```

Comment:
It's good to return more info in errors. I have changed the error guidelines. Pls read the Google error guide

Reply:
which link are you talking about? i saw links on logs for control and data path

Reply:
I will check in